File: Data Structures & Algorithms/validate-parentheses/submission-2.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(' || c =='[' || c == '{') {
                st.push(c);
            } else if (c == ')' && isValid('(', st)){
                return false;
            } else if (c == ']' && isValid('[', st)) {
                return false;
            } else if (c == '}' && isValid('{', st)) {
                return false;
            }
        }
        return st.isEmpty();
    }
    public boolean isValid(char c, Stack<Character> st) {
        if (st.isEmpty()) {
            return true;
        }
        return st.pop() != c;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-2.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        
        stack = []
        answer = 0
        
        for i in range(len(heights)):
            start = i
            while stack and stack[-1][1] > heights[i]:
                index, height = stack.pop()
                answer = max(answer, height * (i - index))
                start = index
            stack.append((start, heights[i]))
                
        while stack:
            index, height = stack.pop()
            answer = max(answer, height * (len(heights) - index))          

        return answer